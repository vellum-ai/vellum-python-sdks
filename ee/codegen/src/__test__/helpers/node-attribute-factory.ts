import { NodeAttribute } from "src/types/vellum";

export const nodeAttributeFactory = (
  id: string,
  name: string,
  attributes: NodeAttribute[]
): NodeAttribute => ({
  id,
  name,
  value: {
    type: "CONSTANT_VALUE",
    value: {
      type: "JSON",
      value: attributes,
    },
  },
});
