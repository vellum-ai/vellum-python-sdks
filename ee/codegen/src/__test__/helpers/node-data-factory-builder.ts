import {
  AdornmentNode,
  CodeResourceDefinition,
  NodeAttribute,
  NodeOutput,
  NodePort,
  NodeTrigger,
  WorkflowDataNode as WorkflowDataNodeType,
} from "src/types/vellum";

export class NodeDataFactoryBuilder<T extends WorkflowDataNodeType> {
  public nodeData: T;

  constructor(nodeData: T) {
    this.nodeData = nodeData;
  }

  withPorts(ports: NodePort[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      ports: ports,
    };
    return this;
  }

  withAttributes(attributes: NodeAttribute[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      attributes,
    };
    return this;
  }

  withAdornments(adornments: AdornmentNode[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      adornments,
    };
    return this;
  }

  withOutputs(outputs: NodeOutput[]): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      outputs,
    };
    return this;
  }

  withBase(base: CodeResourceDefinition): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      base,
    };
    return this;
  }

  withTrigger(trigger: NodeTrigger): NodeDataFactoryBuilder<T> {
    this.nodeData = {
      ...this.nodeData,
      trigger,
    };
    return this;
  }

  build(): T {
    return this.nodeData;
  }
}
