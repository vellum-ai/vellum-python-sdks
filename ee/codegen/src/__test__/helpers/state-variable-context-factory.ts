import { VellumVariable } from "vellum-ai/api/types";

import { WorkflowContext } from "src/context";
import { StateVariableContext } from "src/context/state-variable-context";

export function stateVariableContextFactory({
  stateVariableData,
  workflowContext,
}: {
  stateVariableData: VellumVariable;
  workflowContext: WorkflowContext;
}): StateVariableContext {
  return new StateVariableContext({
    stateVariableData,
    workflowContext,
  });
}
