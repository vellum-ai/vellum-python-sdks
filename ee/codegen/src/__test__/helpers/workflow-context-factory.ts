import { entrypointNodeDataFactory } from "./node-data-factories";

import { WorkflowContext } from "src/context";

export function workflowContextFactory({
  absolutePathToOutputDirectory,
  moduleName,
  workflowClassName,
  workflowRawData,
  strict = true,
  disableFormatting = true,
}: Partial<WorkflowContext.Args> = {}): WorkflowContext {
  return new WorkflowContext({
    absolutePathToOutputDirectory:
      absolutePathToOutputDirectory || "./src/__tests__/",
    moduleName: moduleName || "code",
    workflowClassName: workflowClassName || "TestWorkflow",
    vellumApiKey: "<TEST_API_KEY>",
    workflowRawData: workflowRawData || {
      nodes: [entrypointNodeDataFactory()],
      edges: [],
    },
    strict,
    disableFormatting,
  });
}
