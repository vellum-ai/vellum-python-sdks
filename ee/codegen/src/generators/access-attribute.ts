import { AstNode } from "@fern-api/python-ast/core/AstNode";
import { Writer } from "@fern-api/python-ast/core/Writer";

export declare namespace AccessAttribute {
  interface Args {
    lhs: AstNode;
    rhs: AstNode;
  }
}

export class AccessAttribute extends AstNode {
  private readonly lhs: AstNode;
  private readonly rhs: AstNode;

  constructor({ lhs, rhs }: AccessAttribute.Args) {
    super();

    this.lhs = lhs;
    this.inheritReferences(lhs);

    this.rhs = rhs;
    this.inheritReferences(rhs);
  }

  public write(writer: Writer): void {
    this.lhs.write(writer);
    writer.write(".");
    this.rhs.write(writer);
  }
}
