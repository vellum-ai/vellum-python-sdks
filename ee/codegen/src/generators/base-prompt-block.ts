import { python } from "@fern-api/python-ast";
import { ClassInstantiation } from "@fern-api/python-ast/ClassInstantiation";
import { MethodArgument } from "@fern-api/python-ast/MethodArgument";
import { AstNode } from "@fern-api/python-ast/core/AstNode";
import { Writer } from "@fern-api/python-ast/core/Writer";
import { isNil } from "lodash";
import {
  ChatMessagePromptBlock,
  JinjaPromptBlock,
  PlainTextPromptBlock,
  RichTextPromptBlock,
  VariablePromptBlock,
} from "vellum-ai/api";

import { VELLUM_CLIENT_MODULE_PATH } from "src/constants";
import { WorkflowContext } from "src/context/workflow-context";
import {
  FunctionDefinitionPromptTemplateBlock,
  PlainTextPromptTemplateBlock,
  PromptTemplateBlock,
} from "src/types/vellum";

export type PromptTemplateBlockExcludingFunctionDefinition =
  | Exclude<PromptTemplateBlock, FunctionDefinitionPromptTemplateBlock>
  | PlainTextPromptTemplateBlock;

export type PromptBlock =
  | JinjaPromptBlock
  | ChatMessagePromptBlock
  | VariablePromptBlock
  | PlainTextPromptBlock
  | RichTextPromptBlock;

export declare namespace BasePromptBlock {
  interface Args<
    T extends PromptTemplateBlockExcludingFunctionDefinition | PromptBlock
  > {
    workflowContext: WorkflowContext;
    promptBlock: T;
    inputVariableNameById?: Record<string, string>;
  }
}

export abstract class BasePromptBlock<
  T extends PromptTemplateBlockExcludingFunctionDefinition | PromptBlock
> extends AstNode {
  protected workflowContext: WorkflowContext;
  private astNode: python.ClassInstantiation;
  protected inputVariableNameById: Record<string, string> | undefined; // Stateful prompt blocks have an input variable name by id, prompt blocks do not

  public constructor({
    workflowContext,
    promptBlock,
    inputVariableNameById,
  }: BasePromptBlock.Args<T>) {
    super();
    this.workflowContext = workflowContext;
    this.inputVariableNameById = inputVariableNameById;
    this.astNode = this.generateAstNode(promptBlock);
  }

  protected abstract generateAstNode(promptBlock: T): ClassInstantiation;

  protected constructCommonClassArguments(promptBlock: T): MethodArgument[] {
    const args: MethodArgument[] = [];

    if (promptBlock.state && promptBlock.state !== "ENABLED") {
      args.push(
        new MethodArgument({
          name: "state",
          value: python.TypeInstantiation.str(promptBlock.state),
        })
      );
    }

    const cacheConfig = this.generateCacheConfig(promptBlock);
    if (cacheConfig) {
      args.push(
        new MethodArgument({
          name: "cache_config",
          value: cacheConfig,
        })
      );
    }

    return args;
  }

  private generateCacheConfig(promptBlock: T): python.AstNode | undefined {
    if (isNil(promptBlock.cacheConfig)) {
      return undefined;
    }

    if (!promptBlock.cacheConfig.type) {
      return undefined;
    }

    const cacheConfigType = python.TypeInstantiation.str(
      promptBlock.cacheConfig.type
    );

    return python.instantiateClass({
      classReference: python.reference({
        name: "EphemeralPromptCacheConfig",
        modulePath: VELLUM_CLIENT_MODULE_PATH,
      }),
      arguments_: [
        new MethodArgument({ name: "type", value: cacheConfigType }),
      ],
    });

    return python.TypeInstantiation.none();
  }

  public write(writer: Writer): void {
    this.astNode.write(writer);
  }
}
