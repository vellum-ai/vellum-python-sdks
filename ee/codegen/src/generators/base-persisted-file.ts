import { mkdir, writeFile } from "fs/promises";
import path, { join } from "path";

import { Comment } from "@fern-api/python-ast/Comment";
import { Reference } from "@fern-api/python-ast/Reference";
import { StarImport } from "@fern-api/python-ast/StarImport";
import { AstNode } from "@fern-api/python-ast/core/AstNode";
import { Writer } from "@fern-api/python-ast/core/Writer";
import { ModulePath } from "@fern-api/python-ast/core/types";

import { INIT_FILE_NAME } from "src/constants";
import { WorkflowContext } from "src/context";
import { FileFormattingError } from "src/generators/errors";
import { PythonFile } from "src/generators/extensions";

export declare namespace BasePersistedFile {
  interface Args {
    workflowContext: WorkflowContext;
    isInitFile?: boolean;
  }
}

const VELLUM_MODULES = new Set(["vellum", "vellum_ee"]);

class ImportSortedPythonFile extends PythonFile {
  writeImports({
    writer,
    uniqueReferences,
  }: {
    writer: Writer;
    uniqueReferences: Map<
      string,
      { modulePath: ModulePath; references: Reference[] }
    >;
  }) {
    const sortedReferences = this.sortReferences(uniqueReferences);
    let lastPriority: number | null = null;

    for (const {
      fullyQualifiedPath,
      modulePath,
      references,
      priority,
    } of sortedReferences) {
      const refModulePath = modulePath;
      // Check to see if the reference is defined in this same file and if so, skip its import
      if (this.isDefinedInFile(refModulePath)) {
        continue;
      }

      if (priority !== lastPriority) {
        if (lastPriority !== null) {
          writer.newLine();
        }
        lastPriority = priority;
      }

      if (refModulePath[0] === this.path[0]) {
        // Relativize the import
        // Calculate the common prefix length
        let commonPrefixLength = 0;
        while (
          commonPrefixLength < this.path.length &&
          commonPrefixLength < refModulePath.length &&
          this.path[commonPrefixLength] === refModulePath[commonPrefixLength]
        ) {
          commonPrefixLength++;
        }
        // Calculate the number of levels to go up
        let levelsUp = this.path.length - commonPrefixLength;
        // If this is an __init__.py file, then we must go one more level up.
        if (this.isInitFile) {
          levelsUp++;
        }
        // Build the relative import path
        let relativePath = levelsUp > 0 ? ".".repeat(levelsUp) : ".";
        relativePath += refModulePath.slice(commonPrefixLength).join(".");
        // Write the relative import statement
        writer.write(
          `from ${relativePath} import ${references
            .map((reference) => this.getImportName({ writer, reference }))
            .join(", ")}`
        );
      } else {
        // Use fully qualified path
        writer.write(
          `from ${fullyQualifiedPath} import ${references
            .map((reference) => this.getImportName({ writer, reference }))
            .join(", ")}`
        );
      }
      writer.newLine();
    }
    if (Object.keys(uniqueReferences).length > 0) {
      writer.newLine();
    }
  }

  // The above logic is copied and pasted from the Fern PythonFile AST Library
  // The below is our only contribution to the class
  private sortReferences(
    uniqueReferences: Map<
      string,
      { modulePath: ModulePath; references: Reference[] }
    >
  ) {
    return Array.from(uniqueReferences.entries())
      .map(([fullyQualifiedPath, { modulePath, references }]) => ({
        fullyQualifiedPath,
        modulePath,
        references: references.sort(),
        priority: this.getModulePathPriority(modulePath),
      }))
      .sort((a, b) => {
        if (a.priority !== b.priority) {
          return b.priority - a.priority;
        }
        if (a.references[0]?.name === "*" && b.references[0]?.name !== "*") {
          return 1;
        }
        if (b.references[0]?.name === "*" && a.references[0]?.name !== "*") {
          return -1;
        }
        return a.modulePath.join(".").localeCompare(b.modulePath.join("."));
      });
  }

  private getModulePathPriority(modulePath: ModulePath) {
    if (!modulePath[0]) {
      return 0;
    }

    if (modulePath[0] === this.path[0]) {
      return 1;
    }

    if (VELLUM_MODULES.has(modulePath[0])) {
      return 2;
    }

    return 3;
  }
}

export abstract class BasePersistedFile extends AstNode {
  protected readonly workflowContext: WorkflowContext;
  protected readonly isInitFile: boolean;

  public constructor({ workflowContext, isInitFile }: BasePersistedFile.Args) {
    super();
    this.workflowContext = workflowContext;
    this.isInitFile = isInitFile ?? false;
  }

  protected abstract getModulePath(): string[];

  protected abstract getFileStatements(): AstNode[] | undefined;

  protected getFileImports(): StarImport[] | undefined {
    return undefined;
  }

  protected getComments(): Comment[] | undefined {
    return undefined;
  }

  public write(writer: Writer): void {
    const fileStatements = this.getFileStatements();
    if (!fileStatements) {
      return;
    }

    const file = new ImportSortedPythonFile({
      path: this.getModulePath(),
      isInitFile: this.isInitFile,
      statements: fileStatements,
      imports: this.getFileImports(),
      comments: this.getComments(),
    });

    file.inheritReferences(this);

    file.write(writer);
  }

  public async persist(): Promise<void> {
    const absolutePathToModuleDirectory =
      this.workflowContext.absolutePathToOutputDirectory;

    const modulePath = this.getModulePath();

    let fileName: string;
    let filePath: string[];
    if (this.isInitFile) {
      fileName = INIT_FILE_NAME;
      filePath = modulePath;
    } else {
      fileName = modulePath[modulePath.length - 1] + ".py";
      filePath = modulePath.slice(0, -1);
    }

    const filepath = join(absolutePathToModuleDirectory, ...filePath, fileName);
    await mkdir(path.dirname(filepath), { recursive: true });

    const writer = new Writer();
    this.write(writer);

    let contents: string;
    try {
      contents = await writer.toStringFormatted({ line_width: 120 });
      contents = this.postprocessDocstrings(contents);
    } catch (error) {
      if (error instanceof Error) {
        console.error(`Error formatting ${fileName}:`, error);
      } else {
        console.error(
          new FileFormattingError(
            `Error formatting ${fileName}: ${String(error)}`,
            { cause: error }
          )
        );
      }
      contents = writer.toString();
    }

    await writeFile(filepath, contents);
  }

  private postprocessDocstrings(content: string): string {
    // Find an escaped quote followed by space(s) right before closing triple quotes
    const fixPattern = /(\\")(\s)(""")/g;

    // Remove the extra space between escaped quote and closing triple quotes
    return content.replace(fixPattern, "$1$3");
  }
}
