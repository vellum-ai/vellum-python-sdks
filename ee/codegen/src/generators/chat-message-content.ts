import { python } from "@fern-api/python-ast";
import { MethodArgument } from "@fern-api/python-ast/MethodArgument";
import { AstNode } from "@fern-api/python-ast/core/AstNode";
import { Writer } from "@fern-api/python-ast/core/Writer";
import { isNil } from "lodash";
import {
  ChatMessageContentRequest as ChatMessageContentRequestType,
  ChatMessageContent as ChatMessageContentType,
  FunctionCallChatMessageContentValueRequest as FunctionCallChatMessageContentValueRequestType,
  FunctionCallChatMessageContentValue as FunctionCallChatMessageContentValueType,
  ArrayChatMessageContentItemRequest as ArrayChatMessageContentItemRequestType,
  ArrayChatMessageContentItem as ArrayChatMessageContentItemType,
  VellumImage as VellumImageType,
  VellumImageRequest as VellumImageRequestType,
  VellumAudio as VellumAudioType,
  VellumAudioRequest as VellumAudioRequestType,
  VellumDocument as VellumDocumentType,
  VellumDocumentRequest as VellumDocumentRequestType,
  // VellumVideo as VellumVideoType,
  // VellumVideoRequest as VellumVideoRequestType,
} from "vellum-ai/api";

import { VELLUM_CLIENT_MODULE_PATH } from "src/constants";
import { Json } from "src/generators/json";
import { removeEscapeCharacters } from "src/utils/casing";
import { assertUnreachable } from "src/utils/typing";

class StringChatMessageContent extends AstNode {
  private astNode: AstNode;

  public constructor(value: string, isRequestType: boolean) {
    super();
    this.astNode = this.generateAstNode(value, isRequestType);
  }

  private generateAstNode(value: string, isRequestType: boolean): AstNode {
    const astNode = python.instantiateClass({
      classReference: python.reference({
        name: "StringChatMessageContent" + (isRequestType ? "Request" : ""),
        modulePath: VELLUM_CLIENT_MODULE_PATH,
      }),
      arguments_: [
        python.methodArgument({
          name: "value",
          value: python.TypeInstantiation.str(removeEscapeCharacters(value)),
        }),
      ],
    });
    this.inheritReferences(astNode);
    return astNode;
  }

  public write(writer: Writer): void {
    this.astNode.write(writer);
  }
}

class FunctionCallChatMessageContentValue extends AstNode {
  private astNode: AstNode;

  public constructor(
    value:
      | FunctionCallChatMessageContentValueRequestType
      | FunctionCallChatMessageContentValueType,
    isRequestType: boolean
  ) {
    super();
    this.astNode = this.generateAstNode(value, isRequestType);
  }

  private generateAstNode(
    value:
      | FunctionCallChatMessageContentValueRequestType
      | FunctionCallChatMessageContentValueType,
    isRequestType: boolean
  ): AstNode {
    const functionCallChatMessageContentValueArgs: MethodArgument[] = [];

    if (!isNil(value.id)) {
      functionCallChatMessageContentValueArgs.push(
        new MethodArgument({
          name: "id",
          value: python.TypeInstantiation.str(value.id),
        })
      );
    }

    functionCallChatMessageContentValueArgs.push(
      new MethodArgument({
        name: "name",
        value: python.TypeInstantiation.str(value.name),
      })
    );

    if (value.arguments !== undefined) {
      functionCallChatMessageContentValueArgs.push(
        new MethodArgument({
          name: "arguments",
          value: new Json(value.arguments),
        })
      );
    }

    const functionCallChatMessageContentValueRequestRef = python.reference({
      name:
        "FunctionCallChatMessageContentValue" +
        (isRequestType ? "Request" : ""),
      modulePath: VELLUM_CLIENT_MODULE_PATH,
    });

    const functionCallChatMessageContentValueInstance = python.instantiateClass(
      {
        classReference: functionCallChatMessageContentValueRequestRef,
        arguments_: functionCallChatMessageContentValueArgs,
      }
    );

    const astNode = python.instantiateClass({
      classReference: python.reference({
        name:
          "FunctionCallChatMessageContent" + (isRequestType ? "Request" : ""),
        modulePath: VELLUM_CLIENT_MODULE_PATH,
      }),
      arguments_: [
        new MethodArgument({
          name: "value",
          value: functionCallChatMessageContentValueInstance,
        }),
      ],
    });
    this.inheritReferences(astNode);
    return astNode;
  }

  public write(writer: Writer): void {
    this.astNode.write(writer);
  }
}

class ArrayChatMessageContent extends AstNode {
  private astNode: AstNode;

  public constructor(
    value:
      | ArrayChatMessageContentItemRequestType[]
      | ArrayChatMessageContentItemType[],
    isRequestType: boolean
  ) {
    super();
    this.astNode = this.generateAstNode(value, isRequestType);
  }

  private generateAstNode(
    value:
      | ArrayChatMessageContentItemRequestType[]
      | ArrayChatMessageContentItemType[],
    isRequestType: boolean
  ): AstNode {
    const arrayElements = value.map(
      (element) =>
        new ChatMessageContent({
          chatMessageContent: element as ChatMessageContentRequestType,
        })
    );
    const astNode = python.instantiateClass({
      classReference: python.reference({
        name: "ArrayChatMessageContent" + (isRequestType ? "Request" : ""),
        modulePath: VELLUM_CLIENT_MODULE_PATH,
      }),
      arguments_: [
        python.methodArgument({
          name: "value",
          value: python.TypeInstantiation.list(arrayElements, {
            endWithComma: true,
          }),
        }),
      ],
    });
    this.inheritReferences(astNode);
    return astNode;
  }

  public write(writer: Writer): void {
    this.astNode.write(writer);
  }
}

class AudioChatMessageContent extends AstNode {
  private astNode: AstNode;

  public constructor(
    value: VellumAudioType | VellumAudioRequestType,
    isRequestType: boolean
  ) {
    super();
    this.astNode = this.generateAstNode(value, isRequestType);
  }

  private generateAstNode(
    value: VellumAudioType | VellumAudioRequestType,
    isRequestType: boolean
  ): AstNode {
    const audioChatMessageContentRequestRef = python.reference({
      name: "AudioChatMessageContent" + (isRequestType ? "Request" : ""),
      modulePath: VELLUM_CLIENT_MODULE_PATH,
    });

    const audioArgs = [
      python.methodArgument({
        name: "src",
        value: python.TypeInstantiation.str(value.src),
      }),
    ];

    if (!isNil(value.metadata)) {
      const metadataJson = new Json(value.metadata);
      audioArgs.push(
        python.methodArgument({
          name: "metadata",
          value: metadataJson,
        })
      );
    }

    const arguments_ = [
      python.methodArgument({
        name: "value",
        value: python.instantiateClass({
          classReference: python.reference({
            name: "VellumAudio" + (isRequestType ? "Request" : ""),
            modulePath: VELLUM_CLIENT_MODULE_PATH,
          }),
          arguments_: audioArgs,
        }),
      }),
    ];

    const astNode = python.instantiateClass({
      classReference: audioChatMessageContentRequestRef,
      arguments_: arguments_,
    });
    this.inheritReferences(astNode);
    return astNode;
  }

  public write(writer: Writer): void {
    this.astNode.write(writer);
  }
}

class VideoChatMessageContent extends AstNode {
  private astNode: AstNode;

  public constructor(
    value: unknown, // VellumVideoType | VellumVideoRequestType,
    isRequestType: boolean
  ) {
    super();
    this.astNode = this.generateAstNode(value, isRequestType);
  }

  private generateAstNode(
    value: unknown, // VellumVideoType | VellumVideoRequestType,
    isRequestType: boolean
  ): AstNode {
    const videoChatMessageContentRequestRef = python.reference({
      name: "VideoChatMessageContent" + (isRequestType ? "Request" : ""),
      modulePath: VELLUM_CLIENT_MODULE_PATH,
    });

    const videoArgs = [
      python.methodArgument({
        name: "src",
        // @ts-ignore
        value: python.TypeInstantiation.str(value.src),
      }),
    ];

    // @ts-ignore
    if (!isNil(value.metadata)) {
      // @ts-ignore
      const metadataJson = new Json(value.metadata);
      videoArgs.push(
        python.methodArgument({
          name: "metadata",
          value: metadataJson,
        })
      );
    }

    const arguments_ = [
      python.methodArgument({
        name: "value",
        value: python.instantiateClass({
          classReference: python.reference({
            name: "VellumVideo" + (isRequestType ? "Request" : ""),
            modulePath: VELLUM_CLIENT_MODULE_PATH,
          }),
          arguments_: videoArgs,
        }),
      }),
    ];

    const astNode = python.instantiateClass({
      classReference: videoChatMessageContentRequestRef,
      arguments_: arguments_,
    });
    this.inheritReferences(astNode);
    return astNode;
  }

  public write(writer: Writer): void {
    this.astNode.write(writer);
  }
}

class ImageChatMessageContent extends AstNode {
  private astNode: AstNode;

  public constructor(
    value: VellumImageType | VellumImageRequestType,
    isRequestType: boolean
  ) {
    super();
    this.astNode = this.generateAstNode(value, isRequestType);
  }

  private generateAstNode(
    value: VellumImageType | VellumImageRequestType,
    isRequestType: boolean
  ): AstNode {
    const imageChatMessageContentRequestRef = python.reference({
      name: "ImageChatMessageContent" + (isRequestType ? "Request" : ""),
      modulePath: VELLUM_CLIENT_MODULE_PATH,
    });

    const imageArgs = [
      python.methodArgument({
        name: "src",
        value: python.TypeInstantiation.str(value.src),
      }),
    ];

    if (!isNil(value.metadata)) {
      const metadataJson = new Json(value.metadata);
      imageArgs.push(
        python.methodArgument({
          name: "metadata",
          value: metadataJson,
        })
      );
    }

    const arguments_ = [
      python.methodArgument({
        name: "value",
        value: python.instantiateClass({
          classReference: python.reference({
            name: "VellumImage" + (isRequestType ? "Request" : ""),
            modulePath: VELLUM_CLIENT_MODULE_PATH,
          }),
          arguments_: imageArgs,
        }),
      }),
    ];

    const astNode = python.instantiateClass({
      classReference: imageChatMessageContentRequestRef,
      arguments_: arguments_,
    });
    this.inheritReferences(astNode);
    return astNode;
  }

  public write(writer: Writer): void {
    this.astNode.write(writer);
  }
}

class DocumentChatMessageContent extends AstNode {
  private astNode: AstNode;

  public constructor(
    value: VellumDocumentType | VellumDocumentRequestType,
    isRequestType: boolean
  ) {
    super();
    this.astNode = this.generateAstNode(value, isRequestType);
  }

  private generateAstNode(
    value: VellumDocumentType | VellumDocumentRequestType,
    isRequestType: boolean
  ): AstNode {
    const documentChatMessageContentRequestRef = python.reference({
      name: "DocumentChatMessageContent" + (isRequestType ? "Request" : ""),
      modulePath: VELLUM_CLIENT_MODULE_PATH,
    });

    const documentArgs = [
      python.methodArgument({
        name: "src",
        value: python.TypeInstantiation.str(value.src),
      }),
    ];

    if (!isNil(value.metadata)) {
      const metadataJson = new Json(value.metadata);
      documentArgs.push(
        python.methodArgument({
          name: "metadata",
          value: metadataJson,
        })
      );
    }

    const arguments_ = [
      python.methodArgument({
        name: "value",
        value: python.instantiateClass({
          classReference: python.reference({
            name: "VellumDocument" + (isRequestType ? "Request" : ""),
            modulePath: VELLUM_CLIENT_MODULE_PATH,
          }),
          arguments_: documentArgs,
        }),
      }),
    ];

    const astNode = python.instantiateClass({
      classReference: documentChatMessageContentRequestRef,
      arguments_: arguments_,
    });
    this.inheritReferences(astNode);
    return astNode;
  }

  public write(writer: Writer): void {
    this.astNode.write(writer);
  }
}

export namespace ChatMessageContent {
  export interface Args {
    chatMessageContent: ChatMessageContentRequestType | ChatMessageContentType;
    isRequestType?: boolean;
  }
}

export class ChatMessageContent extends AstNode {
  private astNode: AstNode;

  public constructor({
    chatMessageContent,
    isRequestType = false,
  }: ChatMessageContent.Args) {
    super();
    this.astNode = this.generateAstNode(chatMessageContent, isRequestType);
  }

  private generateAstNode(
    chatMessageContent: ChatMessageContentRequestType | ChatMessageContentType,
    isRequestType: boolean
  ): AstNode {
    let astNode: AstNode;

    const contentType = chatMessageContent.type;
    switch (contentType) {
      case "STRING": {
        astNode = new StringChatMessageContent(
          chatMessageContent.value,
          isRequestType
        );
        break;
      }
      case "FUNCTION_CALL": {
        astNode = new FunctionCallChatMessageContentValue(
          chatMessageContent.value,
          isRequestType
        );
        break;
      }
      case "ARRAY": {
        astNode = new ArrayChatMessageContent(
          chatMessageContent.value,
          isRequestType
        );
        break;
      }
      case "AUDIO": {
        astNode = new AudioChatMessageContent(
          chatMessageContent.value,
          isRequestType
        );
        break;
      }
      // @ts-expect-error
      case "VIDEO": {
        astNode = new VideoChatMessageContent(
          // @ts-expect-error
          chatMessageContent.value,
          isRequestType
        );
        break;
      }
      case "IMAGE": {
        astNode = new ImageChatMessageContent(
          chatMessageContent.value,
          isRequestType
        );
        break;
      }
      case "DOCUMENT": {
        astNode = new DocumentChatMessageContent(
          chatMessageContent.value,
          isRequestType
        );
        break;
      }
      default: {
        assertUnreachable(contentType);
      }
    }

    this.inheritReferences(astNode);
    return astNode;
  }

  public write(writer: Writer): void {
    this.astNode.write(writer);
  }
}
