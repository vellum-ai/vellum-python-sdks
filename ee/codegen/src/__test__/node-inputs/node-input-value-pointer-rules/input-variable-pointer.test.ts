import { Writer } from "@fern-api/python-ast/core/Writer";

import {
  nodeContextFactory,
  workflowContextFactory,
} from "src/__test__/helpers";
import { inputVariableContextFactory } from "src/__test__/helpers/input-variable-context-factory";
import { InputVariablePointerRule } from "src/generators/node-inputs";

describe("InputVariablePointer", () => {
  let writer: Writer;

  beforeEach(() => {
    writer = new Writer();
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it("should generate correct Python code", async () => {
    const workflowContext = workflowContextFactory();
    workflowContext.addInputVariableContext(
      inputVariableContextFactory({
        inputVariableData: {
          id: "test-input-id",
          key: "testVariable",
          type: "STRING",
        },
        workflowContext,
      })
    );

    const nodeContext = await nodeContextFactory({ workflowContext });

    const inputVariablePointer = new InputVariablePointerRule({
      nodeContext,
      nodeInputValuePointerRule: {
        type: "INPUT_VARIABLE",
        data: {
          inputVariableId: "test-input-id",
        },
      },
    });

    inputVariablePointer.write(writer);

    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });

  it("should handle when it's referencing an input variable that no longer exists", async () => {
    const workflowContext = workflowContextFactory();
    const nodeContext = await nodeContextFactory({ workflowContext });

    const inputVariablePointer = new InputVariablePointerRule({
      nodeContext,
      nodeInputValuePointerRule: {
        type: "INPUT_VARIABLE",
        data: {
          inputVariableId: "missing-input-id",
        },
      },
    });

    inputVariablePointer.write(writer);

    expect(await writer.toStringFormatted()).toMatchSnapshot();
  });
});
