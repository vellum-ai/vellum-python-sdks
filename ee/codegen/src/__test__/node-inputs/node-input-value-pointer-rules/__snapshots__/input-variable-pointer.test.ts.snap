// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InputVariablePointer > should generate correct Python code 1`] = `
"Inputs.testVariable
"
`;

exports[`InputVariablePointer > should handle when it's referencing an input variable that no longer exists 1`] = `
"None
"
`;
